gobblin-utility/src/main/java/gobblin/util/AvroFlattener.java

```diff
@@ -0,0 +1,296 @@
+
+    return flattenedSchema;
+  }
+
+  private static Schema flatten(Schema schema, int level) {
+    // New schema
+    Schema flattenedSchema;
+
+    // Process all Schema Types
+    // (Primitives are simply cloned)
+    switch (schema.getType()) {
+      case ARRAY:
+        // Array might be an array of recursive Records, flatten them
+        flattenedSchema = Schema.createArray(flatten(schema.getElementType(), level+1));
+        break;
+      case BOOLEAN:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case BYTES:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case DOUBLE:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case ENUM:
+        flattenedSchema =
+            Schema.createEnum(schema.getName(), schema.getDoc(), schema.getNamespace(), schema.getEnumSymbols());
+        break;
+      case FIXED:
+        flattenedSchema =
+            Schema.createFixed(schema.getName(), schema.getDoc(), schema.getNamespace(), schema.getFixedSize());
+        break;
+      case FLOAT:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case INT:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case LONG:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case MAP:
+        flattenedSchema = Schema.createMap(flatten(schema.getValueType(), level+1));
+        break;
+      case NULL:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case RECORD:
+        List<Schema.Field> flattenedFields = new ArrayList<>();
+        if (null != schema.getFields() && schema.getFields().size() > 0) {
+          // Recursively flatten fields
+          for (Schema.Field oldField : schema.getFields()) {
+            LinkedList<String> lineage = new LinkedList<>();
+            lineage.add(schema.getName());
+            List<Schema.Field> newFields = flattenField(oldField, lineage, level+1, false);
+            if (null != newFields && newFields.size() > 0) {
+              flattenedFields.addAll(newFields);
+            }
+          }
+        }
+        flattenedSchema = Schema.createRecord(schema.getName(), schema.getDoc(), schema.getNamespace(),
+            schema.isError());
+        flattenedSchema.setFields(flattenedFields);
+        break;
+      case STRING:
+        flattenedSchema = Schema.create(schema.getType());
+        break;
+      case UNION:
+        List<Schema> flattenedUnionMembers = new ArrayList<>();
+        if (null != schema.getTypes() && schema.getTypes().size() > 0) {
+          for (Schema oldUnionMember : schema.getTypes()) {
+            // Its member might still recursively contain records
+            flattenedUnionMembers.add(flatten(oldUnionMember, level + 1));
+          }
+        }
+        flattenedSchema = Schema.createUnion(flattenedUnionMembers);
```

Comment:
1. Union of a [record_type, null] and [null, record_type] should be handled as a special case. It should be flattened to the result of the flattening of record_type and all fields should be optional.
2. [a lower priority, might add as an extension in the future] The case of union of [record_type_1, record_type_2, ...] may be handled by combining the flattened record_type_1, record_type_2, ... and using pre-pending the name of the record types to each field name for disambiguation.